### DISCUSSION TITLE
lpcXpresso55s16 adc support

### DISCUSSION THREAD

**Discussion by ArjunDinesh007** [zephyrproject-rtos/zephyr#58309] [Q&A] [answered]
Hi Team, 

Please help me to provide the configuration steps and sample codes of adc read for lpcXpresso55s16.

**Comment by DerekSnell** [comment]
Hi @ArjunDinesh007 ,

The ADC driver is not yet enabled on the LPCXpresso55s16 board yet.  Each NXP board page documents the current [supported features](https://docs.zephyrproject.org/latest/boards/arm/lpcxpresso55s16/doc/index.html#supported-features).  This document also states:

> For additional features not yet supported, please also refer to the [NXP LPCXPRESSO55S69](https://docs.zephyrproject.org/latest/boards/arm/lpcxpresso55s69/doc/index.html#lpcxpresso55s69) , which is the superset board in NXP’s LPC55xx series. NXP prioritizes enabling the superset board with NXP’s Full Platform Support for Zephyr. Therefore, the lpcxpresso55s69 board may have additional features already supported, which can also be re-used on this lpcxpresso55s16 board

The ADC is [supported on the LPCXpresso55s69 board](url), and is the same ADC peripheral and driver on the LPC55S16.  So if you want to add that feature, you can use the LPCXpresso55s69 as a reference.

Best regards

**Reply by ArjunDinesh007** [reply]
Hi DerekSnell,

Thanks for your immediate response. 
I referred the following tutorial and the ADC reference of Zephyr :
[https://www.youtube.com/watch?v=Y8FaEcAp_kY](url) , [https://docs.zephyrproject.org/latest/samples/drivers/adc/README.html](url)

I made changes to the LPCXpresso55s16 board file by referring the board file of  LPCXpresso55s69:

1) Added pinmux_lpadc0 configuration to lpcxpresso55s16-pinctrl.dtsi file
`pinmux_lpadc0: pinmux_lpadc0 {
		group0 {
			pinmux = <ADC0_CH0_PIO0_23>,
				<ADC0_CH8_PIO0_16>,
				<ADC0_CH4_PIO1_8>,
				<ADC0_CH12_PIO1_9>;
			slew-rate = "standard";
			nxp,analog-mode;
		};
	};`

2) Enabled ADC through adding the below code to lpcxpresso55s16.dts file
`&adc0 {
	status = "okay";
	pinctrl-0 = <&pinmux_lpadc0>;
	pinctrl-names = "default";
};`

3) Added 'adc' under supported section of the lpcxpresso55s16.yaml file
`identifier: lpcxpresso55s16
name: NXP LPCXpresso55S16
type: mcu
arch: arm
ram: 96
flash: 256
toolchain:
  - zephyr
  - gnuarmemb
  - xtools
supported:
  - adc
  - arduino_gpio
  - arduino_i2c
  - arduino_spi
  - can
  - gpio
  - i2c
  - spi
  - usb_device`

4) Added adc0 under the aliases field of the lpcxpresso55s16_common.dtsi file
`aliases{
		led0 = &red_led;
		led1 = &green_led;
		led2 = &blue_led;
		sw0 = &btn_wk;
		sw1 = &btn_usr;
		sw2 = &btn_isp;
		adcctrl = &adc0;
		usart-0 = &flexcomm0;
		magn0 = &fxos8700;
		accel0 = &fxos8700;
	};`

I used the below code to read the adc samples from LPCXpresso55s16 through PIO0_23/ADC0_0 (12th pin - LPC55S16JBD64)
`#include <stdint.h>
#include <stdio.h>
#include <zephyr.h>
#include <device.h>
#include <drivers/adc.h>
#include "logging/osal_log.h"
#include <can_com.h>
#include "can/osal_can.h"

#define ADC_DEVICE_NAME "DT_LABEL(DT_ALIAS(adcctrl))"
OSAL_LOG_MODULE_REGISTER(com_data_handler);

void main(void)
{
    enum adc_gain {
        ADC_GAIN_1_6,
        ADC_GAIN_1_5,
        ADC_GAIN_1_4,
        ADC_GAIN_1_3,
        ADC_GAIN_1_2,
        ADC_GAIN_2_3,
        ADC_GAIN_1,
        ADC_GAIN_2,
        ADC_GAIN_3, 
        ADC_GAIN_4,
        ADC_GAIN_8, 
        ADC_GAIN_16,
        ADC_GAIN_32,
        ADC_GAIN_64,
    };

    enum adc_reference {
        ADC_REF_VDD_1,
        ADC_REF_VDD_1_2,
        ADC_REF_VDD_1_3,
        ADC_REF_VDD_1_4,
        ADC_REF_INTERNAL,
        ADC_REF_EXTERNAL0,
        ADC_REF_EXTERNAL1,
    };

    enum adc_input_positive {
        ADC_POSITIVE_INPUT_AIN0,
        ADC_POSITIVE_INPUT_AIN1,
        ADC_POSITIVE_INPUT_AIN2,
        ADC_POSITIVE_INPUT_AIN3,
        ADC_POSITIVE_INPUT_AIN4,
        ADC_POSITIVE_INPUT_AIN5,
        ADC_POSITIVE_INPUT_AIN6,
        ADC_POSITIVE_INPUT_AIN7,
        ADC_POSITIVE_INPUT_AIN8,
        ADC_POSITIVE_INPUT_AIN9,
        ADC_POSITIVE_INPUT_AIN10,
        ADC_POSITIVE_INPUT_AIN11,
        ADC_POSITIVE_INPUT_AIN12,
        ADC_POSITIVE_INPUT_AIN13,
        ADC_POSITIVE_INPUT_AIN14,
        ADC_POSITIVE_INPUT_AIN15,
    };

    struct adc_channel_cfg {
        uint32_t acquisition_time;
        uint8_t channel_id;
        enum adc_input_positive input_positive;
        enum adc_gain gain;
        enum adc_reference reference;
    };

    struct adc_sequence {
        uint16_t *buffer;
        uint16_t buffer_size;
        uint8_t resolution;
        uint8_t channels;
        struct adc_channel_cfg *channels_cfg;
        const struct adc_sequence_options *options;
        uint32_t timeout;
        bool calibrate;
    };

    const struct device *adc_dev;
    struct adc_channel_cfg channel_cfg;
    struct adc_sequence sequence;
    uint16_t buffer[4];

    adc_dev = device_get_binding(ADC_DEVICE_NAME);
    if (!adc_dev) {
        OSAL_LOG_INF("Failed to get ADC device\n");
        return;
    }

    channel_cfg.gain = ADC_GAIN_1_6;
    channel_cfg.reference = ADC_REF_INTERNAL;
    channel_cfg.acquisition_time = ADC_ACQ_TIME(ADC_ACQ_TIME_MICROSECONDS, 40);
    channel_cfg.channel_id = 0;
    channel_cfg.input_positive = ADC_POSITIVE_INPUT_AIN0;

    sequence.buffer = buffer;
    sequence.buffer_size = sizeof(buffer);
    sequence.resolution = 16;
    sequence.channels = 1;

    sequence.options = NULL;
    sequence.timeout = SYS_FOREVER_MS;
    sequence.calibrate = false;
    sequence.channels_cfg = &channel_cfg;

    while (1) {
        if (adc_read(adc_dev, &sequence) != 0) {
            OSAL_LOG_INF("ADC read failed\n");
            return;
        }

        OSAL_LOG_INF("ADC value: %u\n", buffer[0]);

        k_sleep(K_SECONDS(1));
    }
}
`
I am able to build the code without any error. But, after connecting the real LPCXpresso55s16 board and flashed the hex file into the board.
The output is showing error:

*** Booting Zephyr OS build zephyr-v3.3.0-3441-g3154e03c454a ***
[00:00:00.005,000] <inf> com_data_handler: Failed to get ADC device

I am not sure where I made mistakes. If there is any possibilities, please guide me to resolve this issue.
Any suggestions and corrections are greatly appreciated.

Thanks in advance.

**Comment by DerekSnell** [comment]
Hi @ArjunDinesh007 ,
This looks like a devicetree issue getting the `adcctrl` device in this line:
```
#define ADC_DEVICE_NAME "DT_LABEL(DT_ALIAS(adcctrl))"
```

I am not a devicetree expert, and typically follow the samples to reference the devicetree nodes.  One example you can reference is the [ADC driver sample](https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/drivers/adc).  Or if you prefer to use an alias like this, the [blink ](https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/basic/blinky)and [button ](https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/basic/button)samples use alias nodes.

Best regards

**Reply by ArjunDinesh007** [reply]
Hi @DerekSnell,

First of all thank you for the support and suggestions.

Regarding the device tree configuration, I already added the adc0 under the aliases field of the lpcxpresso55s16_common.dtsi file
`aliases{ adcctrl = &adc0;};`
It is not showing any build errors.

As per the adc driver sample example, I saw the overlay file for the LPCXpresso55s69. Since the adc pin configurations of the LPCXpresso55s69 and LPCXpresso55s16 is matching, I tried to copy the overlay file of the LPCXpresso55s69 and changed the file name to LPCXpresso55s16.

I am able to build the code without any error. But, after connecting the real LPCXpresso55s16 board and flashed the hex file into the board.
The output is showing the same error:

*** Booting Zephyr OS build zephyr-v3.3.0-3441-g3154e03c454a ***
[00:00:00.005,000] com_data_handler: Failed to get ADC device

I am not sure whether I followed the correct steps or not. Please advice me.

Thanking you,
Arjun D

**Reply by DerekSnell** [reply]
Hi @ArjunDinesh007 ,
I see you edited your response, and still seem to be having issues with the app accessing the ADC node from the devicetree.  It seems you are testing with a custom application here, right?  To rule out an issue with that app, have you confirmed if you app also runs on the lpcxpresso55s69 board?  Or have you confirmed the [ADC driver sample](https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/drivers/adc) is able to run on the lpcxpresso55s16 board?
Best regards

**Reply by ArjunDinesh007** [reply]
Hi @DerekSnell ,

I agree with you. It seems to be the device tree error.
I went to more trials based on this reference : https://docs.zephyrproject.org/latest/hardware/porting/board_porting.html

I referred the ‘board porting guide’, Made changes to the adc configuration of lpc55s16, which I missed in the previous trials:

1) Added the configuration "select HAS_MCUX_LPADC" under - zephyr\soc\arm\nxp_lpc\lpc55xxx\Kconfig.soc as per the adc soc configuration mentioned in the lpc55s69
`
config SOC_LPC55S69_CPU0
            bool "SOC_LPC55S69 M33 [CPU 0]"
            select CPU_CORTEX_M33
            select CPU_HAS_ARM_SAU
            select CPU_HAS_ARM_MPU
            select CPU_HAS_FPU
            select ARMV8_M_DSP
            select ARM_TRUSTZONE_M
            select CLOCK_CONTROL
            select HAS_MCUX_IAP
            select HAS_MCUX_LPADC
            select HAS_MCUX_LPC_DMA
            select HAS_MCUX_USB_LPCIP3511
            select HAS_MCUX_CTIMER
            select HAS_MCUX_SCTIMER
            select HAS_MCUX_RNG
`
`
config SOC_LPC55S16
            bool "SOC_LPC55S16 M33"
            select CPU_CORTEX_M33
            select CPU_HAS_ARM_SAU
            select CPU_HAS_ARM_MPU
            select CPU_HAS_FPU
            select ARMV8_M_DSP
            select ARM_TRUSTZONE_M
            select CLOCK_CONTROL
            select HAS_MCUX_IAP if !TRUSTED_EXECUTION_NONSECURE
            select HAS_MCUX_LPADC
            select HAS_MCUX_MCAN
            select HAS_MCUX_RNG
`
2) Copied the overlay file available for lpc55s69 and taken a copy of the same and renamed it to lpc55s16 (because the adc pin config for lpc55s69 and lpc55s16 is matching) and saved it under - zephyr\samples\drivers\adc\boards

I got a build error "parse error: undefined node label 'adc0'".

By adding the below configuration under lpc55s16.dtsi file, this error got fixed.
adc0: adc@400A0000 {
                        compatible = "nxp,lpc55xx-adc";
                        reg = <0x400A0000 0x1000>;
            };
Note : The reg address is updated based on the datasheet of lpc55s16(0x400A0000 - ADC0)

Now, I am able to build the code without errors. 
After flashing the code to the real LPCXpresso55s16 board, it is showing the same error:

*** Booting Zephyr OS build zephyr-v3.3.0-3441-g3154e03c454a ***
Failed to get ADC device

Please support with your valuable suggestions.

**Comment by DerekSnell** [comment]
Hi @ArjunDinesh007 ,

I am still waiting for you to confirm that you are testing this with an app that we know works.  Please confirm:
Does the [ADC driver sample](https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/drivers/adc) work correctly on the lpcxpresso55s16 board?

or

Does this app you are testing work correctly on the lpcxpresso55s69 board?

Thanks

**Reply by ArjunDinesh007** [reply]
Hi @DerekSnell,

We are working with a custom application and the aim is to read the voltage value from the sensor through the ADC channel of lpcxpresso55s16 board.

I tried with the [ADC driver sample](https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/drivers/adc) code as you suggested,

Made changes to: zephyr\samples\drivers\adc\sample.yaml file:
1) included the board name 'lpcxpresso55s16' to sample.yaml file under "platform_allow:" & "integration_platforms:" fields.
    (Please find the attached sample.yaml file)

[sample.zip](https://github.com/zephyrproject-rtos/zephyr/files/11732094/sample.zip)

2) Used a copied version of the overlay file of lpcxpresso55s69_cpu0.overlay and renamed it to - lpcxpresso55s16.overlay (Placed it under zephyr\samples\drivers\adc\boards)

While building the code (with the command: west build -t run -b lpcxpresso55s16), I got build errors:

1) 

`
devicetree error: C:/ENDOR/zephyrproject/zephyr/samples/drivers/adc/boards/lpcxpresso55s16.overlay:42 (column 30): parse error: expected number or parenthesized expression
devicetree error: C:/ENDOR/zephyrproject/zephyr/samples/drivers/adc/boards/lpcxpresso55s16.overlay:57 (column 30): parse error: expected number or parenthesized expression
devicetree error: C:/ENDOR/zephyrproject/zephyr/samples/drivers/adc/boards/lpcxpresso55s16.overlay:71 (column 30): parse error: expected number or parenthesized expression
`

Resolved this error by changing the corresponding lines inside the lpcxpresso55s16.overlay file (line numbers: 42, 57, 71 as mentioned above) by adding the below edits:

`zephyr,acquisition-time = "ADC_ACQ_TIME(ADC_ACQ_TIME_MICROSECONDS, 20)";
`

2)

`
-- Found devicetree overlay: C:/ENDOR/zephyrproject/zephyr/samples/drivers/adc/boards/lpcxpresso55s16.overlay
devicetree error: /zephyr,user: undefined node label 'adc'
-- In: C:/ENDOR/zephyrproject/zephyr/samples/drivers/adc/build/zephyr, command: C:/ENDOR/zephyrproject/.venv/Scripts/python.exe;C:/ENDOR/zephyrproject/zephyr/scripts/dts/gen_defines.py;--dts;C:/ENDOR/zephyrproject/zephyr/samples/drivers/adc/build/zephyr/zephyr.dts.pre;--dtc-flags;'-Wno-simple_bus_reg -Wno-unique_unit_address_if_enabled';--bindings-dirs;C:/ENDOR/zephyrproject/zephyr/dts/bindings;--header-out;C:/ENDOR/zephyrproject/zephyr/samples/drivers/adc/build/zephyr/include/generated/devicetree_generated.h.new;--dts-out;C:/ENDOR/zephyrproject/zephyr/samples/drivers/adc/build/zephyr/zephyr.dts.new;--edt-pickle-out;C:/ENDOR/zephyrproject/zephyr/samples/drivers/adc/build/zephyr/edt.pickle;--vendor-prefixes;C:/ENDOR/zephyrproject/zephyr/dts/bindings/vendor-prefixes.txt
CMake Error at C:/ENDOR/zephyrproject/zephyr/cmake/modules/dts.cmake:261 (message):
  gen_defines.py failed with return code: 1
Call Stack (most recent call first):
  C:/ENDOR/zephyrproject/zephyr/cmake/modules/zephyr_default.cmake:115 (include)
  C:/ENDOR/zephyrproject/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:66 (include)
  C:/ENDOR/zephyrproject/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:97 (include_boilerplate)
  CMakeLists.txt:5 (find_package)`

`
-- Configuring incomplete, errors occurred!
`

Resolved this error by using the below edits:
`#include <zephyr/dt-bindings/adc/mcux-lpadc.h>
/ {
	zephyr,user {
		io-channels = <&adc0>;
	};
};
`

After edited this configuration changes, I tried building the sample code.
Now, It is showing the warning and error as shown below:

`
C:\ENDOR\zephyrproject\zephyr\samples\drivers\adc>west build -t run -b lpcxpresso55s16
-- west build: running target run
[0/1] Re-running CMake...Loading Zephyr default modules (Zephyr base (cached)).
-- Application: C:/ENDOR/zephyrproject/zephyr/samples/drivers/adc
-- CMake version: 3.26.3
-- Cache files will be written to: C:/ENDOR/zephyrproject/zephyr/.cache
-- Zephyr version: 3.3.99 (C:/ENDOR/zephyrproject/zephyr)
-- Found west (found suitable version "1.0.0", minimum required is "0.7.1")
-- Board: lpcxpresso55s16
-- Found host-tools: zephyr 0.15.2 (C:/ENDOR/zephyrproject/zephyr-sdk-0.15.2)
-- Found toolchain: zephyr 0.15.2 (C:/ENDOR/zephyrproject/zephyr-sdk-0.15.2)
-- Found BOARD.dts: C:/ENDOR/zephyrproject/zephyr/boards/arm/lpcxpresso55s16/lpcxpresso55s16.dts
-- Found devicetree overlay: C:/ENDOR/zephyrproject/zephyr/samples/drivers/adc/boards/lpcxpresso55s16.overlay
-- Generated zephyr.dts: C:/ENDOR/zephyrproject/zephyr/samples/drivers/adc/build/zephyr/zephyr.dts
-- Generated devicetree_generated.h: C:/ENDOR/zephyrproject/zephyr/samples/drivers/adc/build/zephyr/include/generated/devicetree_generated.h
-- Including generated dts.cmake file: C:/ENDOR/zephyrproject/zephyr/samples/drivers/adc/build/zephyr/dts.cmake
C:/ENDOR/zephyrproject/zephyr/samples/drivers/adc/build/zephyr/zephyr.dts:411.3-27: Warning (io_channels_property): /zephyr,user:io-channels: property size (4) too small for cell size 1
Parsing C:/ENDOR/zephyrproject/zephyr/Kconfig
Loaded configuration 'C:/ENDOR/zephyrproject/zephyr/samples/drivers/adc/build/zephyr/.config'
Configuration saved to 'C:/ENDOR/zephyrproject/zephyr/samples/drivers/adc/build/zephyr/.config'
Kconfig header saved to 'C:/ENDOR/zephyrproject/zephyr/samples/drivers/adc/build/zephyr/include/generated/autoconf.h'
Load components for LPC55S16:
driver_common component is included.
driver_reset component is included.
device_CMSIS component is included.
CMSIS_Include_core_cm component is included.
device_system component is included.
driver_lpc_gpio component is included.
driver_pint component is included.
driver_inputmux component is included.
driver_inputmux_connections component is included.
driver_flexcomm_usart component is included.
driver_flexcomm component is included.
driver_lpc_iocon component is included.
-- Configuring done (21.5s)
-- Generating done (2.0s)
-- Build files have been written to: C:/ENDOR/zephyrproject/zephyr/samples/drivers/adc/build`
`
[0/1] cmd.exe /C "cd /D C:\ENDOR\zephyrproject\zephyr\samples\drivers\adc\build\zephyr && "C:\Program Fi...mulation/Simulation not supported with this board." ==================================================="[1/1] cmd.exe /C "cd /D C:\ENDOR\zephyrproject\zephyr\samples\drivers\adc\build\zephyr && "C:\Program Fi...mulation/Simulation not supported with this board." ==================================================="←[K
=================================================== Emulation/Simulation not supported with this board. ===================================================
`

I am hereby sharing with you the board files (lpcxpresso55s16) and the overlay file which we are using for this testing. Please find the attached files.
[lpcxpresso55s16.zip](https://github.com/zephyrproject-rtos/zephyr/files/11732130/lpcxpresso55s16.zip)

[lpcxpresso55s16_overlay.zip](https://github.com/zephyrproject-rtos/zephyr/files/11732220/lpcxpresso55s16_overlay.zip)

Addressing your second query: The lpcxpresso55s69 board is not available with us. So, I am unable to test and verify this sample code with lpcxpresso55s69 at this moment.

Please support us.

**Comment by DerekSnell** [comment, marked as answer]
Hi @ArjunDinesh007 ,
First, the `-t run` argument in your `west build` command does not work for these boards.  And this is causing your build error "Emulation/Simulation not supported with this board."  This argument is used for for emulation on host systems like QEMU.  See [Zephyr's Building doc page](https://docs.zephyrproject.org/latest/develop/west/build-flash-debug.html#building-west-build) for more details.  

To build for the `lpcxpresso55s69` board, you can use this command below.  I realize you do not have an LPCXpresso55S69 board available (BTW, I would recommend getting one.  It will help confirm the working operation on that board, as you port new features to the board you are using), but since this ADC driver sample does work on the `lpcxpresso55s69` board, you can build the sample for that board to compare with the build for your target board.
```
west build samples\drivers\adc -b lpcxpresso55s69_cpu0
```

Also, your error log shows an issue in your devicetree:
```
C:/ENDOR/zephyrproject/zephyr/samples/drivers/adc/build/zephyr/zephyr.dts:411.3-27: Warning (io_channels_property): /zephyr,user:io-channels: property size (4) too small for cell size 1
```

The file zephyr.dts is generated during the build, and shows the final merged devicetree that is compiled with the application.  You can review line 411 in this zephyr.dts file to find the node and property generating this warning.  You can also build the same sample application for the `lpcxpresso55s69` board.  And then compare these nodes in the zephyr.dts file generated for that board, to see how they differ.  Since you are porting this ADC driver feature from the `lpcxpresso55s69` board, comparing these zephyr.dts files may show something missing from your ported devicetree.

I have not yet reviewed the files you shared.  If I have some time, I will take a look at your port.  In the meantime, let us know what you find after resolving these issues.  Best regards

**Reply by ArjunDinesh007** [reply]
Hi @DerekSnell,

Still I am blocked with the lpcxpresso55s69 testing because the board is not available with me. I requested support from NXP team for the same based on your suggestions. Meanwhile did you get a chance to reviewing the files which I shared? If you reviewed the files, then please share your valuable feedback. Best regards
